crates/rspack_plugin_javascript_shared/src/js_parser_hook.rs

```diff
@@ -0,0 +1,160 @@
+use std::{borrow::Cow, collections::HashMap};
+
+use rspack_core::{CodeGeneratableDependency, ModuleDependency};
+use swc_core::ecma::{
+  ast,
+  visit::{Visit, VisitWith},
+};
+
+type StaticStr = Cow<'static, str>;
+
+#[derive(Debug, Default)]
+pub struct JsParserContext {
+  pub presentational_dependencies: Vec<Box<dyn CodeGeneratableDependency>>,
+  pub dependencies: Vec<Box<dyn ModuleDependency>>,
+}
+
+/// The reason why we introduce [JsParserHookDriver] instead of using plain `swc_core::ecma_visit::Visit` is
+/// we want to finish all work in a single visit pass.
+#[derive(Default)]
+pub struct JsParserHookDriver {
+  hooks: HashMap<StaticStr, Vec<Box<dyn JsParserHook>>>,
+  context: JsParserContext,
+}
+
+impl JsParserHookDriver {
+  pub fn register_with_key(&mut self, key: impl Into<StaticStr>, hook: Box<dyn JsParserHook>) {
+    let key = key.into();
+    self.hooks.entry(key).or_default().push(hook);
+  }
+
+  pub fn into_context(self) -> JsParserContext {
+    self.context
+  }
+}
+
+pub enum Control {
+  // Skip subsequent subscribers
+  Skip,
+  Continue,
+}
+
+/// This macro is used to help of reducing boilerplate code of adding
+/// hook for various JS AST node type.
+///
+/// # Explanation
+/// For input
+/// ```ignore
+/// register_hooks!(pub trait JsParserHook {
+///   (visit_expr, ast::Expr),
+/// });
+/// ```
+///
+/// The macro would expand as
+///
+/// ```ignore
+/// trait JsParserHook {
+///   fn visit_expr(&mut self, _ctx: &mut JsParserContext, _node: &ast::Expr) {
+///      false
+///   }
+/// }
+///
+/// impl<'me> Visit for JsAstVisitor<'me> {
+///   fn visit_expr(&mut self, node: &ast::Expr) {
+///     self.hooks.values_mut().for_each(|hooks| {
+///       // Bailout if any hook returns `true`
+///       hooks.iter_mut().any(|hook| hook.visit_expr(&self.context, node));
+///     });
+///     node.visit_children_with(self)
+///   }
+/// }
+/// ```
+macro_rules! register_hooks {
+  (
+    pub trait JsParserHook {
+        $( ($name:ident,  $node:ty),)*
+    }
+  ) => {
+    pub trait JsParserHook {
+      $(
+        fn $name(&mut self, _ctx: &mut JsParserContext,_node: &$node) -> Control {
+          Control::Continue
+        }
+      )*
+    }
+
+    impl Visit for JsParserHookDriver {
+        $(
+        fn $name(&mut self, node: &$node) {
+          self.hooks.values_mut().for_each(|hooks| {
+            // Bailout if any hook returns `true`
+            hooks.iter_mut().any(|hook| matches!(hook.$name(&mut self.context, node), Control::Skip));
```

Comment:
This actually looks pretty weird to me. Why would the hook affect other hooks behavior for __analyzation__ itself? Could you give an example for the real-world use case? If this is for performance reason, I think this will make problem hard to discover in the future.

Reply:
That's how webpack works. 

For example: 
https://github.com/webpack/webpack/blob/f1ef2eded2fda8136535c60813dcb0c55510a5e0/lib/APIPlugin.js#L190

Return `true` would skip all the subscribers under the `__webpack_layer__` key.

The real-world example would be `Worker support`. If you have interest in it, we could setup a lark meeting.

Reply:
This looks good to me.